### ISSUE TITLE
Knex migrations are not portable between SQLite3 and other databases

### ISSUE THREAD

**Issue by Kryten0807** [tgriesser/knex#2781]
# Environment

Knex version: 0.14.2
Database + version: SQLite3 3.22.0
Node: 8.11.1
OS: Ubuntu 18.04

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

Knex migrations are not portable between SQLite3 and other databases.

If I want to use Knex in a portable way, I need to be able to write my code such that it will work against a variety of databases without custom code for each database. It should work in my testing environment, where I'm using SQLite3 with a `:memory:` database or against my production MariaDB database.

The example `knexfile.js` and migration below fail on SQLite3 with the error given below. The error message below shows that Knex is generating an SQL statement which is invalid for SQLite3 (which does not work with tablenames that are qualified with a scheme name like `scheme.table`).

2. Error message

    Knex:warning - migration failed with error: create table `mydb`.`mytable` (`id` integer not null primary key autoincrement, `name` varchar(255) not null) - SQLITE_ERROR: unknown database `mydb`
    Error: SQLITE_ERROR: unknown database `mydb`

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to mysql or postgresql, then single file example which initializes
   needed data and demonstrates the problem.

I attempted to set it up on runkit, but even the basic Knex example at the URL above fails with missing dependencies.

    /* knexfile.js */

    const migrations = {
        tableName: 'api_migrations',
    };

    // Update with your config settings.

    module.exports = {

        test: {
            client:           'sqlite3',
            connection:       { filename: ':memory:' },
            useNullAsDefault: true,
        },

        production: {
            client:     'mysql2',
            connection: {
                host:     process.env.DB_HOST,
                user:     process.env.DB_USER,
                password: process.env.DB_PASSWORD,
                database: process.env.DB_DATABASE,
            },
        },

    };

    /* migration: 2018-08-22094726_test.js */
    exports.up = (knex, Promise) =>
        knex.schema.withSchema('mydb').hasTable('mytable').then((exists) =>
            exists
            || knex.schema.withSchema('mydb').createTable('mytable', (table) => {
                table.increments('id');
                table.string('name').notNullable().index();
            })
        );

    exports.down = (knex, Promise) =>
        knex.schema.hasTable('mytable').then((exists) =>
            exists
            || knex.schema.withSchema('mydb').dropTableIfExists('mytable')
        );

**Comment by ricardograca** [comment]
So, sqlite doesn't really support schemas. What are you expecting to happen here?

**Comment by Kryten0807** [comment]
It should probably map from the fully-qualified `scheme`.`table` to something that SQLite3 can handle.

For example, I've written a little helper function that checks the database driver (which is defined in an environment variable) and rearranges the scheme & table names:

    const tableName = (scheme, table) => [scheme, table].join(process.env.DB_CLIENT === 'sqlite3' ? '_' : '.');

If the database is `sqlite3`, this returns `scheme_table`; if it's anything else, it returns `scheme.table`.

This works for SQLite3, but breaks migrations on MySQL/MariaDB because passing a fully-qualified `scheme.table` name to `hasTable` always returns `false`.

There are probably multiple ways to implement this. My thought is something like the following: when using the `sqlite3` driver, a query using the `withSchema()` method will use the schema and table name to build an SQLite-friendly name (perhaps `scheme_name`).

---

Perhaps this is a problem with my expectations.

The [Knex website](https://knexjs.org/) says this package is "designed to be flexible, portable, and fun to use".

I understand "portable" to mean that Knex aims to provide a common API to the database systems it supports. In other words, I can write my code once, without reference to my chosen database engine, and then I can switch database engine simply by changing the configuration in my knexfile.

Is this the intention? If it is, then it seems to me that this issue identifies a situation where that portability is broken.

**Comment by ricardograca** [comment]
> I understand "portable" to mean that Knex aims to provide a common API to the database systems it supports.
Speaking only for myself I think that's the general idea, although it's not always possible, like in this case. One thing that could happen in this case is that knex simply ignores the schema part. Not sure if that would be better or worse than the current situation.

**Comment by Kryten0807** [comment]
I would argue that this *is* a bug. Consider the following examples from the documentation:

    knex('books')
      .returning(['id', 'title'])
      .insert({title: 'Slaughterhouse Five'})

On postgresql, it returns `id` & `title`; on MySQL (and other databases that don't support it), it returns simply the new `id` value. It doesn't generate invalid SQL and the difference in behaviour is clearly documented.

And this example:

    knex.schema.table('mytable', (table) => {
        table.integer('my_new_column').after('id');
    });

On MySQL it inserts the new column after the `id` column; on other databases that don't support this, it is ignored. It doesn't generate invalid SQL for other database engines and the difference in behaviour is mentioned in the documentation.

Contrast those two examples with the following:

    knex.withSchema('public').select('*').from('users')

The documentation clearly states that this generates SQL with the schema and table, with both delimited by backticks. No mention of different behaviour on different database engines (I assume it properly delimits names on MSSQL?). And it attempts to execute an invalid SQL statement against SQLite.

In short, there are some database features (`returning`, `after`) which are not universally supported by database engines but are permitted without breaking anything. They are handled in a sensible way (IMO) on databases that do not support them and they are documented.

We have another feature (`withSchema`) that is not universally supported by database engines. In this case, `knex` attempts to run invalid SQL against the database and fails.

One sensible and (hopefully) not to difficult-to-implement way to handle this would be to follow the suggestion @ricardograca made: simply ignore the schema. This could cause problems for some people, so it would probably be good to emit a warning (much like the warning message for the `useNullAsDefault` setting).

**Comment by elhigu** [comment]
Its better that code throws an error than just working plainly wrong. The same applies for returning and other, which actually doesnt throw yet, but only creates invalid code which actually doesn’t work the same way with different dialects. 

Behaviour where code silently works wrong is way more dangerous, than having an exception of not supported db feature.

You will have false sense of safety thinking that hey this code is portable when it actually does completely different things with different databases, which one should handed in the user side of the library. 

I’ve already discussed about this theme many times before, but I dont have laptop to look up related discussions right now.
